### PR TITLE
[BE][doc] add memory_profiler to README

### PR DESCRIPTION
**Pull request by mori360**
Add [memory_profiler](https://github.com/pytorch/torchtitan/pull/395) to README, explain how to use memory profiler with `--profiling.enable_memory_snapshot` and `--profiling.save_memory_snapshot_folder`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +42,7 @@ You may want to see how the model is defined or how parallelism techniques are a
 10. DDP and HSDP
 11. All options easily configured via [toml files](train_configs/)
 12. [Interoperable checkpoints](docs/checkpoint.md) which can be loaded directly into [`torchtune`](https://github.com/pytorch/torchtune) for fine-tuning
+13. [Memory profier](docs/memory_profiler.md) dump memory snapshots
```
Comment (weifengpy) [inline README.md]: maybe modify ine 38?  "6. Loss, GPU memory, tokens-per-second, and MFU displayed and logged via [TensorBoard](https://github.com/pytorch/torchtitan#tensorboard)" ? 

this is a debugging feature instead of a major PTD feature

File: docs/memory_profiler.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+## Enable Memory Profiling
+
+Launch training job with the following command (or alternatively set configs in toml files)
+```
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder output_folder
+```
+* `--profiling.enable_memory_snapshot`: enable memory snapshot
+* `--profiling.save_memory_snapshot_folder`: dump memory snapshots in to output folder, default to be `memory_snapshot`.
+	+ If in case of OOMs. output folder is `memory_snapshot/iteration_x_exit`.
+	+ If regularly according to `profile_freq`. output folder is `memory_snapshot/iteration_x`.
```
Comment (weifengpy) [inline docs/memory_profiler.md]: explain how people should open the pickle file? like go to browser, drag, and check. we should quote https://pytorch.org/blog/understanding-gpu-memory-1/

File: docs/memory_profiler.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+## Enable Memory Profiling
+
+Launch training job with the following command (or alternatively set configs in toml files)
+```
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder output_folder
+```
+* `--profiling.enable_memory_snapshot`: enable memory snapshot
+* `--profiling.save_memory_snapshot_folder`: dump memory snapshots in to output folder, default to be `memory_snapshot`.
```
Comment (tianyu-l) [inline docs/memory_profiler.md]: The default here should be `./outputs/memory_snapshot`

File: README.md (hunk 1 above)
Comment (tianyu-l) [reply README.md]: We should add an item in the Debugging section together with Flight Recorder. Personally I don't mind creating a new entry in "key features" about debugging tools, but they should be both included or both excluded here.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: docs/memory_profiler.md (hunk 3 above)
Comment (tianyu-l) [reply docs/memory_profiler.md]: I mean the default value for this field is `memory_snapshot`, but you should mention it would be under the output folder, which is controlled by `job.dump_folder`.

File: README.md (hunk 1 above)
Comment (mori360) [reply README.md]: Combine them together in line 45 as bullet point 13.

File: docs/memory_profiler.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+## Enable Memory Profiling
+
+Launch training job with the following command (or alternatively set configs in toml files)
+```
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder output_folder
+```
+* `--profiling.enable_memory_snapshot`: enable memory snapshot
+* `--profiling.save_memory_snapshot_folder`: dump memory snapshots in to output folder, default under your output folder to be `./outputs/memory_snapshot`.
+	+ If in case of OOMs. output folder is `memory_snapshot/iteration_x_exit`.
```
Comment (fegin) [inline docs/memory_profiler.md]: The first "." should be replace with ",".

File: docs/memory_profiler.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+	+ If regularly according to `profile_freq`. output folder is `memory_snapshot/iteration_x`.
```
Comment (fegin) [inline docs/memory_profiler.md]: ditto.

File: README.md [hunk 6]
```diff
@@ -42,6 +42,7 @@ You may want to see how the model is defined or how parallelism techniques are a
 10. DDP and HSDP
 11. All options easily configured via [toml files](train_configs/)
 12. [Interoperable checkpoints](docs/checkpoint.md) which can be loaded directly into [`torchtune`](https://github.com/pytorch/torchtune) for fine-tuning
+13. CPU/GPU profiling, [memory profiling](docs/memory_profiler.md), [flight recorder](#debugging)
```
Comment (tianyu-l) [inline README.md]:
```suggestion
13. Debugging tools including CPU/GPU profiling, [memory profiling](docs/memory_profiler.md), [Flight Recorder](#debugging), etc.
```

File: docs/memory_profiler.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+## Enable Memory Profiling
+
+Launch training job with the following command (or alternatively set configs in toml files)
+```
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder output_folder
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder memory_snapshot
```

File: docs/memory_profiler.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+```
+* `--profiling.enable_memory_snapshot`: enable memory snapshot
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
* `--profiling.enable_memory_snapshot`: to enable memory profiling
```

File: docs/memory_profiler.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+* `--profiling.save_memory_snapshot_folder`: dump memory snapshots in to output folder, default under your output folder to be `./outputs/memory_snapshot`.
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
* `--profiling.save_memory_snapshot_folder`: configures the folder which memory snapshots are dumped into (`./outputs/memory_snapshot/` by default)
```

File: docs/memory_profiler.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+	+ If in case of OOMs, output folder is `memory_snapshot/iteration_x_exit`.
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
	+ In case of OOMs, the snapshots will be in `./outputs/memory_snapshot/iteration_x_exit`.
```

File: docs/memory_profiler.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+	+ If regularly according to `profile_freq`, output folder is `memory_snapshot/iteration_x`.
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
	+ Regular snapshots (taken every `profiling.profile_freq` iterations) will be in `memory_snapshot/iteration_x`.
```

File: docs/memory_profiler.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+Once you have dumped the memory profiler, you will find the saved pickle files in your output folder.
+To visualize the snapshot file, you can utilize the `memory_viz` tool, by either dragging and dropping the snapshot into your browser or generating its HTML file, following the [tutorial](https://pytorch.org/blog/understanding-gpu-memory-1/).
```
Comment (tianyu-l) [inline docs/memory_profiler.md]:
```suggestion
To visualize a snapshot file, you can drag and drop it to <https://pytorch.org/memory_viz>. To learn more details on memory profiling, please visit this [tutorial](https://pytorch.org/blog/understanding-gpu-memory-1/).
```

**Review by tianyu-l — COMMENTED** [review]
Had some suggestions on writing

File: docs/memory_profiler.md [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+Launch training job with the following command (or alternatively set configs in toml files)
+```
+CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --profiling.enable_memory_snapshot --profiling.save_memory_snapshot_folder memory_snapshot
+```
+* `--profiling.enable_memory_snapshot`: to enable memory profiling
+* `--profiling.save_memory_snapshot_folder`: configures the folder which memory snapshots are dumped into (`./outputs/memory_snapshot/` by default)
+	+ In case of OOMs, the snapshots will be in `./outputs/memory_snapshot/iteration_x_exit`.
+	+ Regular snapshots (taken every `profiling.profile_freq` iterations) will be in `memory_snapshot/iteration_x`.
+
+Once you have dumped the memory profiler, you will find the saved pickle files in your output folder.
```
Comment (tianyu-l) [inline docs/memory_profiler.md]: Let's remove this sentence, as it is redundant.

**Review by tianyu-l — APPROVED** [review]
looks good to me, thanks!
